Add support for private fonts in Direct2D renderer

a-wi opened the pull request:
Since it appears it's not possible to create DirectWrite fonts with IDWriteGdiInterop::CreateFontFromLOGFONT() for fonts registered with AddFontResourceEx() as private ones (with FR_PRIVATE flag), we need to maintain a custom DirectWrite font collection and use it as source of font data if call to CreateFontFromLOGFONT() fails. This collection consists of IDWriteFontFileEnumerator and IDWriteFontCollectionLoader objects which are seeded with font data based on the private fonts list taken from wxGetPrivateFontFileNames().

Closes #18687.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -360,6 +360,221 @@ HRESULT WINAPI wxD3D11CreateDevice(
+
+namespace
+{
+    IDWriteFontCollection* gs_pPrivateFontCollection = NULL;
```
Would it be possible to use `wxCOMPtr` for this one too to avoid explicit `Release()` calls?

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -360,6 +360,221 @@ HRESULT WINAPI wxD3D11CreateDevice(
+extern const wxArrayString& wxGetPrivateFontFileNames();
+
+namespace
+{
```
Very minor, of course, but we typically don't indent the namespaces content, to avoid too deep indentation levels.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -360,6 +360,221 @@ HRESULT WINAPI wxD3D11CreateDevice(
+
+        ~wxDirect2DFontFileEnumerator()
+        {
+        }
```
```suggestion
        // Default dtor is ok.
```
I don't see why should we need to define it.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -360,6 +360,221 @@ HRESULT WINAPI wxD3D11CreateDevice(
+    class wxDirect2DFontFileEnumerator : public IDWriteFontFileEnumerator
+    {
+    public:
+        wxDirect2DFontFileEnumerator(IDWriteFactory* pFactory)
```
```suggestion
        explicit wxDirect2DFontFileEnumerator(IDWriteFactory* pFactory)
```
Just to avoid any strange conversions.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -360,6 +360,221 @@ HRESULT WINAPI wxD3D11CreateDevice(
+        {
+        }
+
+        HRESULT Initialize(wxArrayString& collection)
```
```suggestion
        HRESULT Initialize(const wxArrayString& collection)
```
AFAICS we don't need to modify it.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -360,6 +360,221 @@ HRESULT WINAPI wxD3D11CreateDevice(
+        // IUnknown methods
+        virtual HRESULT STDMETHODCALLTYPE QueryInterface(REFIID iid, void** ppvObject) wxOVERRIDE
+        {
+            if ( IsEqualIID(iid, IID_IUnknown) || IsEqualIID(iid, __uuidof(IDWriteFontFileEnumerator)) )
```
We avoid using `__uuidof` if possible, as it makes the code incompatible with MinGW. Couldn't we just use `IID_IDWriteFontFileEnumerator` here instead?

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -360,6 +360,221 @@ HRESULT WINAPI wxD3D11CreateDevice(
+        {
+        }
+
+        // IUnknown methods
```
Could also use the usual `{DECLARE,IMPLEMENT}_IUNKNOWN_METHODS` macros from `wx/msw/ole/comimpl.h`...

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -360,6 +360,221 @@ HRESULT WINAPI wxD3D11CreateDevice(
+            }
+            *ppFontFile = m_currentFile;
+
+            return (m_currentFile != NULL) ? S_OK : E_FAIL;
```
Extremely minor, but:
```suggestion
            return m_currentFile ? S_OK : E_FAIL;
```
`!= NULL` is inconsistent with omitting it just above and the parentheses are just unnecessary.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -360,6 +360,221 @@ HRESULT WINAPI wxD3D11CreateDevice(
+            if ( key != m_key )
+                return E_INVALIDARG;
+
+            HRESULT hr = pEnumerator->Initialize(m_fontList);
```
Is there any reason to separate `Initialize()` from ctor? I'd pass `m_fontList` directory to the ctor and check that it's not empty here, IMHO it would slightly simplify the code.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -360,6 +360,221 @@ HRESULT WINAPI wxD3D11CreateDevice(
+        static IDWriteFontCollectionLoader* m_pInstance;
+        static wxArrayString m_fontList;
+        static wxDirect2DFontKey m_key;
+    };
```
Those should use `ms_` prefix according to our conventions.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -2843,19 +3073,101 @@ wxD2DFontData::wxD2DFontData(wxGraphicsRenderer* renderer, const wxFont& font, c
+            wxFAIL_MSG(wxString::Format("Failed to create font '%s'", logfont.lfFaceName));
+            return;
+        }
+        // Update font collection if the lsi of private fonts has changed.
```
```suggestion
        // Update font collection if the list of private fonts has changed.
```

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -2843,19 +3073,101 @@ wxD2DFontData::wxD2DFontData(wxGraphicsRenderer* renderer, const wxFont& font, c
+        const wxArrayString& privateFonts = wxGetPrivateFontFileNames();
+        if ( privateFonts.empty() )
+        {
+            wxFAIL_MSG(wxString::Format("Failed to create font '%s'", logfont.lfFaceName));
```
I don't think font creation failure is really supposed to result in an assert. `wxLogError()` would probably be more appropriate, but I didn't check what did the other implementations do.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -2843,19 +3073,101 @@ wxD2DFontData::wxD2DFontData(wxGraphicsRenderer* renderer, const wxFont& font, c
+        wxCHECK_RET(gs_pPrivateFontCollection != NULL, "No custom font collection created");
+
+        UINT32 fontIdx;
+        BOOL fontFound;
```
Don't we get warnings about using possible uninitialized variables if we don't initialize them here?

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -2843,19 +3073,101 @@ wxD2DFontData::wxD2DFontData(wxGraphicsRenderer* renderer, const wxFont& font, c
+        hr = gs_pPrivateFontCollection->GetFontFamily(fontIdx, &fontFamily);
+        wxCHECK_HRESULT_RET(hr);
+
+        DWRITE_FONT_WEIGHT fWeight;
```
I think we can just do
```suggestion
        DWRITE_FONT_WEIGHT fWeight = static_cast<DWRITE_FONT_WEIGHT>(logfont.lfWeight);
```
Even though it's an enum, it clearly should be possible to use any value here.

vadz reviewed (commented):
Thanks for fixing this!

Globally I really don't have anything to say, but there a few possible minor improvements here and there at least some of which are arguably worth making.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -360,6 +369,172 @@ HRESULT WINAPI wxD3D11CreateDevice(
+        ms_isInitialized = true;
+    }
+
+#ifdef __WXDEBUG__
```
Could you please add a comment explaining why we do want to have this empty dtor in debug builds? It's just so unusual that it seems anybody reading this code would have this question, so it would be nice to have an answer to it too.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -423,6 +598,14 @@ IDWriteFactory* wxDWriteFactory()
+        HRESULT hr = gs_IDWriteFactory->RegisterFontCollectionLoader(wxDirect2DFontCollectionLoader::GetLoader());
+        if ( FAILED(hr) )
+        {
+            wxLogError("Could not register custom DirectWrite font loader.");
```
Sorry for continued nitpicking with error handling, but this should be
```suggestion
            wxLogError(_("Could not register custom DirectWrite font loader."));
```
as it's a user-visible string.

vadz commented on src/msw/graphicsd2d.cpp:
```diff
@@ -2843,19 +3026,77 @@ wxD2DFontData::wxD2DFontData(wxGraphicsRenderer* renderer, const wxFont& font, c
+        const wxArrayString& privateFonts = wxGetPrivateFontFileNames();
+        if ( privateFonts.empty() )
+        {
+            wxLogError("Failed to create font '%s' (HRESULT = %x)", logfont.lfFaceName, hr);
```
This should either use `_()` too or be changed to use `wxLogApiError()` (if we don't want to show this message to the user, and we probably don't).

vadz reviewed (commented):
Thanks for the fixes! Just a couple of small points about the error handling remaining now.

a-wi committed:
Add support for private fonts in Direct2D renderer

Since it appears it's not possible to create DirectWrite fonts with IDWriteGdiInterop::CreateFontFromLOGFONT() for fonts registered with AddFontResourceEx() as private ones (with FR_PRIVATE flag), we need to maintain a custom DirectWrite font collection and use it as source of font data if call to CreateFontFromLOGFONT() fails. This collection consists of IDWriteFontFileEnumerator and IDWriteFontCollectionLoader objects which are seeded with font data based on the private fonts list taken from wxGetPrivateFontFileNames().

Closes #18687.

vadz commented:
Thanks again!
